tooling/metamodel-generator/src/main/java/org/hibernate/processor/annotation/LifecycleMethod.java

```diff
@@ -334,27 +354,84 @@ private boolean isGeneratedIdUpsert() {
 	private void preamble(StringBuilder declaration) {
 		declaration
 				.append("\n@Override\npublic ")
+				.append(parameterTypeBopunds())
 				.append(returnType())
 				.append(' ')
 				.append(methodName)
 				.append('(');
 		notNull(declaration);
+		final var parameters = method.getParameters();
+		assert parameters.size() == 1;
+		final VariableElement element = parameters.get(0);
 		declaration
-				.append(annotationMetaEntity.importType(entity))
+				.append(resolveAsString(element.asType()))
 				.append(' ')
-				.append(parameterName)
+				.append(element.getSimpleName())
 				.append(')')
 				.append(" {\n");
 	}
 
+	private String parameterTypeBopunds() {
+		if ( method.getTypeParameters().isEmpty() ) {
+			return "";
+		}
+		return method.getTypeParameters().stream()
+				.map( this::resolveTypeParameter )
+				.collect( joining( ", ", " <", "> " ) );
+	}
+
+	private String resolveAsString(TypeMirror type) {
+		if ( type.getKind().isPrimitive() || type.getKind() == VOID ) {
+			return type.toString();
+		}
+		else if ( type instanceof DeclaredType declaredType ) {
+			final var element = annotationMetaEntity.importType(
+					((TypeElement) declaredType.asElement()).getQualifiedName().toString()
+			);
+			if ( declaredType.getTypeArguments().isEmpty() ) {
+				return element;
+			}
+			return element + declaredType.getTypeArguments().stream().map( this::resolveAsString )
+					.collect( joining( ",", "<", ">" ) );
+		}
+		else if ( type instanceof TypeVariable typeVariable ) {
+			final var value = typeVariable.toString();
+			if ( methodTypeParameters.contains( value ) ) {
+				return value;
+			}
+			return annotationMetaEntity.importType( resolveTypeName( element, method.getEnclosingElement(), value ) );
+		}
+		else if ( type instanceof WildcardType wildcardType ) {
+			return "?"
+				+ (wildcardType.getExtendsBound() == null ? ""
+					: " extends " + resolveAsString( wildcardType.getExtendsBound() ))
+				+ (wildcardType.getSuperBound() == null ? ""
+					: " super " + resolveAsString( wildcardType.getExtendsBound() ));
+		}
+		else if ( type instanceof ArrayType arrayType ) {
+			return resolveAsString( arrayType.getComponentType() ) + "[]";
+		}
+		else {
+			System.err.printf( "Type kind %s - %s%n", type.getKind(), type );
+			throw new UnsupportedOperationException();
```

Comment:
@cigaly This is failing the forbidden APIs check.

Actually, we should not ever throw exceptions from the Processor. At most we should add a message.

Reply:
Apart from that issue, this looks good, @cigaly.

Reply:
Only missing `TypeMirror` that can possibly (although I am not sure how exactly) occur is `IntersectionType`, so I will implement that branch, and for default simply return `type.toString()`

Reply:
@gavinking Should I squash last two commits?

Reply:
> Should I squash last two commits?

Yes, please.